eth2near/utilities/src/eth_proof_generator.rs

```diff
@@ -0,0 +1,178 @@
+use crate::{
+    eth_rpc_client::EthRPCClient,
+    primitives::*,
+    types::{BlockHeader, Log, TransactionReceipt},
+};
+use cita_trie::{MemoryDB, PatriciaTrie, Trie, TrieError};
+use hasher::HasherKeccak;
+use rlp::RlpStream;
+use std::error::Error;
+use std::sync::Arc;
+
+#[derive(Debug)]
+pub struct Proof {
+    pub log_index: u64,
+    pub log_entry_data: Vec<u8>,
+    pub receipt_index: u64,
+    pub receipt_data: Vec<u8>,
+    pub header_data: Vec<u8>,
+    pub proof: Vec<Vec<u8>>,
+}
+
+pub fn get_proof_for_event(tx_hash: &str, log_index: u64, node_url: &str,) -> Result<Proof, Box<dyn Error>> {
+    let client = EthRPCClient::new(node_url);
+
+    let tx_bytes: [u8; 32] = hex::decode(&tx_hash[2..])
+        .expect("Invalid hex string")
+        .try_into()
+        .expect("Invalid hex string");
+
+    let receipt = client.get_transaction_receipt_by_hash(&FixedBytes(tx_bytes))?;
+    let block_header = client.get_block_by_number(receipt.block_number)?;
+    let block_receipts = client.get_block_receipts(receipt.block_number)?;
+
+    let trie = build_receipt_trie(&block_receipts)?;
+
+    let receipt_key = rlp::encode(&receipt.transaction_index);
+    let proof = trie.get_proof(&receipt_key)?;
+
+    let mut log_data: Option<Vec<u8>> = None;
+    for (_, log) in receipt.logs.iter().enumerate() {
+        if log.log_index == log_index {
+            log_data = Some(encode_log(log));
+        }
+    }
+
+    Ok(Proof {
+        log_index,
+        log_entry_data: log_data.expect("Log not found"),
+        receipt_index: receipt.transaction_index,
+        receipt_data: encode_receipt(&receipt),
+        header_data: encode_header(&block_header),
+        proof,
+    })
+}
+
+fn build_receipt_trie(receipts: &[TransactionReceipt],) -> Result<PatriciaTrie<MemoryDB, HasherKeccak>, TrieError> {
+    let memdb = Arc::new(MemoryDB::new(true));
+    let hasher = Arc::new(HasherKeccak::new());
+    let mut trie = PatriciaTrie::new(Arc::clone(&memdb), Arc::clone(&hasher));
```

Comment:
```suggestion
    let mut trie = PatriciaTrie::new(memdb, hasher);
```
Extra clone

Reply:
Fixed in c70cc73eb4cd78f8dfdfb2f1be34a6807cc5c921